File: cpp/leetcode/1157. online-majority-element-in-subarray.cpp
#include "public.h"

class MajorityChecker {
private:
	vector<int> v;
public:
	MajorityChecker(vector<int>& arr) {
		v = arr;
	}

	int query(int left, int right, int threshold) {
		//unordered_map<int, int> um;

		int ans;
		int flag = 0;
		for (int i = left; i <= right; i++)
		{
			if (flag > 0)
			{
				if (ans == v[i])
					flag++;
				else
					flag--;
			}
			else
			{
				flag = 1;
				ans = v[i];
			}
		}

		int num = 0;
		for (int i = left; i <= right; i++)
		{
			if (ans == v[i])
				num++;   //验证该数字是否超过一半
		}

		if (num >= threshold) return ans;
		else return -1;
	}
};

/**
 * Your MajorityChecker object will be instantiated and called as such:
 * MajorityChecker* obj = new MajorityChecker(arr);
 * int param_1 = obj->query(left,right,threshold);
 */

File: javascript/1157. online-majority-element-in-subarray.js
/*
暴力, O(nm), n:arr大小, m:查询次数
不建议
4128ms, 6.67%
*/

/**
 * @param {number[]} arr
 */
var MajorityChecker = function(arr) {
    this.m_arr = arr;
};

/** 
 * @param {number} left 
 * @param {number} right 
 * @param {number} threshold
 * @return {number}
 */
MajorityChecker.prototype.query = function(left, right, threshold) {
    var ans = this.m_arr[left];
	var flag = 1;
	for (var i = left + 1; i <= right; i++)
	{
		if (flag > 0)
		{
			if (ans == this.m_arr[i])
				flag++;
			else
				flag--;
		}
		else
		{
			flag = 1;
			ans = this.m_arr[i];
		}
	}
	
	var num = 0;
	for (var i = left; i <= right; i++)
	{
		if (ans == this.m_arr[i])
			num++;
	}
	
	if (num >= threshold) return ans;
	else return -1;
};

/** 
 * Your MajorityChecker object will be instantiated and called as such:
 * var obj = new MajorityChecker(arr)
 * var param_1 = obj.query(left,right,threshold)
 */